basesandbox/src/ipc/mod.rs

```diff
@@ -0,0 +1,92 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pub mod domain_socket;
+pub mod intra;
+pub mod semaphore;
+use std::collections::hash_map::HashMap;
+
+/// Inter-process obejct that is constructed / synchronized by another Inter-process mechanism.
+pub trait InterProcessUnit {
+    /// Constructs itself with an opaque data that would have been transported by some IPC
+    fn new(data: Vec<u8>) -> Self;
+}
+
+/// One possible abstraction for a server-client style of `Linker`
+pub trait TwoWayInitialize {
+    type Server: InterProcessUnit;
+    type Client: InterProcessUnit;
+
+    fn new(name: String) -> Self;
+    /// Generated two configurations
+    /// which will be feeded to InterProcessUnit::new(),
+    /// for both two ends in two different processes.
+    fn create(&self) -> (Vec<u8>, Vec<u8>);
+}
+
+/// Some implementors of InterProcessUnit would take a single 'path'
+/// since it's quite a trivial model that inter-process objects try to acquire an access to a
+/// inter-process handle that is identified by a system-wide name.
+pub fn init_data_from_path(path: String) -> Vec<u8> {
+    let mut config: HashMap<String, String> = HashMap::new();
+    config.insert("path".to_owned(), path);
+    serde_cbor::to_vec(&config).unwrap()
+}
+
+/// Most of IPC depends on a system-wide name, which looks quite vulnerable for
+/// possible attack. Rather, generating a random name would be more secure.
+pub fn generate_random_name() -> String {
+    let pid = std::process::id();
+    let time = std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).unwrap();
+    let hash = ccrypto::blake256(format!("{:?}{}", time, pid));
+    format!("{:?}", hash)[0..16].to_string()
+}
+
+pub trait IpcSend: Send {
+    /// It might block until counterparty's recv(). Even if not, the order is still guaranteed.
+    fn send(&self, data: &[u8]);
+}
+
+#[derive(Debug, PartialEq)]
+pub enum RecvFlag {
+    TimeOut,
+    Termination,
+}
+
+pub trait Terminate: Send {
+    /// Wake up block on recv with a special flag
+    fn terminate(&self);
+}
+
+pub trait IpcRecv: Send {
+    type MyTerminate: Terminate;
+
+    /// Returns Err only for the timeout or termination wake-up(otherwise panic)
+    fn recv(&self, timeout: Option<std::time::Duration>) -> Result<Vec<u8>, RecvFlag>;
+    /// Create a terminate switch that can be sent to another thread
+    fn create_terminate(&self) -> Self::MyTerminate;
+}
+
+pub trait Ipc: InterProcessUnit + IpcSend + IpcRecv {
+    type SendOnly: IpcSend;
+    type RecvOnly: IpcRecv;
+    // split itself into Send-only and Recv-only. This is helpful for a threading
+    fn split(self) -> (Self::SendOnly, Self::RecvOnly);
```

Comment:
If all `Ipc`s are always splitable into their sender half and receiver half, then it seems more natural just to have getters for them. Also it seems natural to return `Arc`. If not, it's conceptually confusing since an IPC primitive is likely to have links to the halves.

Reply:
What do you mean by `getter`? You mean re-combine?

Reply:
Note that it takes `self`, not `&self`.